[DRA] Fix unexpected errors on ResourceClaim strategy update

sunya-ch opened the pull request:
- Fix unexpected error (immutable field, constraint validation)

<!--  Thanks for sending a pull request!  Here are some tips for you:

1. If this is your first time, please read our contributor guidelines: https://git.k8s.io/community/contributors/guide/first-contribution.md#your-first-contribution and developer guide https://git.k8s.io/community/contributors/devel/development.md#development-guide
2. Please label this pull request according to what type of issue you are addressing, especially if this is a release targeted pull request. For reference on required PR/issue labels, read here:
https://git.k8s.io/community/contributors/devel/sig-release/release.md#issuepr-kind-label
3. Ensure you have added or ran the appropriate tests for your PR: https://git.k8s.io/community/contributors/devel/sig-testing/testing.md
4. If you want *faster* PR reviews, read how: https://git.k8s.io/community/contributors/guide/pull-requests.md#best-practices-for-faster-reviews
5. If the PR is unfinished, see how to mark it: https://git.k8s.io/community/contributors/guide/pull-requests.md#marking-unfinished-pull-requests
-->

#### What type of PR is this?

<!--
Add one of the following kinds:
/kind bug
/kind cleanup
/kind documentation
/kind feature

Optionally add one or more of the following kinds if applicable:
/kind api-change
/kind deprecation
/kind failing-test
/kind flake
/kind regression
-->

/kind bug

#### What this PR does / why we need it:

    - Fix setting distinctAttribute=nil when DRAConsumableCapacity is disabled
    - Add must have no error check when no error expected in strategy test
    - Update test case that causes unexpected errors


~~This PR corrects the matchesAttribute of distinctAttributeConstraint to use value instead of pointer and remove the following missing drop field implementation test in ResourceClaim strategy on DeviceBindingConditions: drop-fields-binding-conditions, drop-fields-binding-conditions-disable-feature-gate, drop-fields-binding-conditions-disable-binding-conditions-feature-gate~~ -> No longer valid, rebased.

#### Which issue(s) this PR is related to:
<!--
Please link relevant issues to help with tracking.

To automatically close the linked issue(s) when this PR is merged,
add the word "Fixes" before the issue number or link.
Do not use "Fixes" if the PR is of kind `failing-test` or `flake`.

Reference KEPs when applicable in addition to specific issues.

Examples:
Fixes #<issue number>
<issue link> (issue in a different repository)
KEP: https://github.com/kubernetes/enhancements/issues/<kep-issue-number>

If there is no associated issue, then write "N/A".
-->

Fixes https://github.com/kubernetes/kubernetes/issues/134961

#### Special notes for your reviewer:


#### Does this PR introduce a user-facing change?
<!--
If no, just write "NONE" in the release-note block below.
If yes, a release note is required:
Enter your extended release note in the block below. If the PR requires additional action from users switching to the new release, include the string "action required".

For more information on release notes see: https://git.k8s.io/community/contributors/guide/release-notes.md
-->
```release-note
- Fix setting distinctAttribute=nil when DRAConsumableCapacity is disabled
```

#### Additional documentation e.g., KEPs (Kubernetes Enhancement Proposals), usage docs, etc.:

<!--
This section can be blank if this pull request does not require a release note.

When adding links which point to resources within git repositories, like
KEPs or supporting documentation, please reference a specific commit and avoid
linking directly to the master branch. This ensures that links reference a
specific point in time, rather than a document that may change over time.

See here for guidance on getting permanent links to files: https://help.github.com/en/articles/getting-permanent-links-to-files

Please use the following format for linking documentation:
- [KEP]: <link>
- [Usage]: <link>
- [Other doc]: <link>
-->
```docs

```

pohly commented on pkg/registry/resource/resourceclaimtemplate/strategy_test.go:
```diff
@@ -475,15 +477,19 @@ func TestClaimTemplateStrategyCreate(t *testing.T) {
 			if errs := strategy.Validate(ctx, obj); len(errs) != 0 {
+				if tc.expectValidationError == "" {
+					t.Fatalf("unexpected error: %v", errs[0])
+				}
```
This is what triggered this whole change. I agree that it is a gap.

All of the `pkg/registry/resource/*/strategy_test.go` files have it (copy-and-paste...). I checked some other API groups. At least the ones I looked at were fine because they don't have test cases with validation errors and thus always expect no errors.

Can you also fix the other `pkg/registry/resource/*/strategy_test.go` files?

sunya-ch replied:
@pohly Sure. But could you please help point me out the line? In devicetaintrule, I don't see the same issue where we pass the error message contains empty message condition when there is an unexpected error.

sunya-ch commented on pkg/registry/resource/resourceclaim/strategy.go:
```diff
@@ -237,6 +237,7 @@ func dropDisabledStatusFields(newClaim, oldClaim *resource.ResourceClaim) {
 	dropDisabledDRAResourceClaimDeviceStatusFields(newClaim, oldClaim)
 	dropDisabledDRAAdminAccessStatusFields(newClaim, oldClaim)
 	dropDisabledDRAResourceClaimConsumableCapacityStatusFields(newClaim, oldClaim)
+	dropDeviceBindingConditionsFields(newClaim, oldClaim)
```
@pohly Sure. I separated here: https://github.com/kubernetes/kubernetes/pull/134964. Will rebase on it once it merges.

pohly replied:
Which test case case is failing? What is it meant to ensure? Sounds to me like it wants to verify behavior that wasn't implemented, in which case you can remove it here and add it back in https://github.com/kubernetes/kubernetes/pull/134964.

sunya-ch replied:
@pohly I have removed three test cases (also added to the description) from resourceclaim test.

```
drop-fields-binding-conditions
drop-fields-binding-conditions-disable-feature-gate
drop-fields-binding-conditions-disable-binding-conditions-feature-gate
```

Now this PR is independent from https://github.com/kubernetes/kubernetes/pull/134964.

cc: @KobayashiD27

mortent reviewed (commented):
Agree with the comments from Patrick. Just a few other comments. Thanks for fixing this.

sunya-ch commented on pkg/api/resourceclaimspec/util.go:
```diff
@@ -135,9 +135,16 @@ func dropDisabledDRAResourceClaimConsumableCapacityFields(new, old *resource.Res
+	var newConstraints []resource.DeviceConstraint
+	for _, constraint := range new.Devices.Constraints {
+		// Add only MatchAttribute (Drop DistinctAttribute)
+		if constraint.MatchAttribute != nil {
```
@mortent It is failing on validation that if constraints in not nil, either matchAttribute or distinctAttribute must be set so I cannot just set distinctAttribute (it will say matchAttribute need to be set).

sunya-ch replied:
@pohly Do I need to move this fix to a separate PR too?

pohly replied:
Maybe. It all depends on how you would describe it - is it really part of "Fix strategy test with unexpected error" (i.e. a fix in the test code) or is it fixing some Kubernetes feature?

sunya-ch replied:
@pohly This one is only for fixing strategy test. Can I keep it here?

sunya-ch replied:
@pohly I'm sorry. It is not only for fixing the test. It is also fixing the issue when we need to drop the field when the feature is disabled too. I have updated the description to `[DRA] Fix unexpected validation error on ResourceClaim strategy update` to cover the update fix. Is it acceptable?

mortent replied:
So I think the most important change in this PR is the change in this file. And the problem isn't a validation error, but that the `DistinctAttribute` field did not get dropped when the Consumable Capacity feature gate is disabled. So I would describe that in the title.

The description is a bit hard to follow, so could we update the language a bit? Also, since this fixes a but, we should add a proper release note.

Thanks for clarifying my original question here. I understand why the current logic is needed.

sunya-ch replied:
@mortent  Thank you. I have updated the description in the code, the PR commit to cover this point, and the release note.

pohly replied:
"Fix removing out DistinctAttribute constraints when DRAConsumableCapacity feature is disabled"

is the point I was looking for. Resolved as far as I am concerned.

liggitt replied:
We don't remove items from lists when dropping fields. If the item is not valid without that attribute, then a validation error disallowing the request with that invalid item is expected.

The goal of the field dropping is to make the API server behave as much as possible like an API server prior to introduction of the field, and erroring on the invalid item is what a previous version API server would do.

liggitt replied:
The code prior to this PR looks correct to me. A validation error when submitting an object with an item in a list which is invalid without the gate-disabled field is expected.

sunya-ch replied:
@liggitt  Thank you. I have updated to set the distinctAttribute to nil and expect validation error instead. I also separate the test case dropping with and without the distinctAttribute.

sunya-ch commented on pkg/registry/resource/resourceclaim/strategy_test.go:
```diff
@@ -630,6 +651,9 @@ func TestStrategyCreate(t *testing.T) {
 			if errs := strategy.Validate(ctx, obj); len(errs) != 0 {
+				if tc.expectValidationError == "" {
+					t.Fatalf("unexpected error: %v", errs[0])
+				}
```
@pohly I tried applying it but it seems that they are a couple of test cases that have more than one error occurs.

For example, 
```
        	Error:      	"[metadata.namespace: Invalid value: "default": field is immutable spec: Invalid value: {"Spec":{"Devices":{"Requests":[{"Name":"req-0","Exactly":null,"FirstAvailable":[{"Name":"subreq-0","DeviceClassName":"class","Selectors":null,"AllocationMode":"ExactCount","Count":1,"Tolerations":null,"Capacity":null}]}],"Constraints":null,"Config":null}}}: field is immutable]" should have 1 item(s), but has 2
        	Test:       	TestStrategyUpdate/keep-fields-prioritized-list
        	Messages:   	exactly one error expected
--- FAIL: TestStrategyUpdate/keep-fields-prioritized-list (0.00s)
```

Also, for
- TestStrategyUpdate/drop-fields-prioritized-list (has 3)
-  TestStatusStrategyUpdate/keep-fields-admin-access-NonAdminNamespace (has 2)

The extra one can be fixed though. I will try.

- metadata.namespace: Invalid value: "default" -->  I can think of two ways: changing default to kube-system (as obj) or set namespace default on oldObj. Do you have preference?
- [spec.spec.devices.requests[0]: Required value: exactly one of `exactly` or `firstAvailable` is required

sunya-ch replied:
I have changed the expected error to []string instead for resourceclaimtemplate updatestrategy check.

sunya-ch committed:
[DRA] Fix unexpected errors on ResourceClaim strategy update

- Fix setting distinctAttribute=nil when DRAConsumableCapacity is disabled
- Add must have no error check when no error expected in strategy test
- Update test case that causes unexpected errors

Signed-off-by: Sunyanan Choochotkaew <sunyanan.choochotkaew1@ibm.com>

sunya-ch commented:
/retest

pohly commented on pkg/registry/resource/resourceclaim/strategy_test.go:
```diff
@@ -152,39 +152,6 @@ var objWithAdminAccessInNonAdminNamespace = &resource.ResourceClaim{
 	},
 }
 
-var objStatusInNonAdminNamespace = &resource.ResourceClaim{
```
No preference from me either way.

mortent reviewed (commented):
Other than the title and the description of the PR, this looks good to me.

pohly reviewed (approved):
/lgtm

pohly commented on pkg/registry/resource/resourceclaim/strategy_test.go:
```diff
@@ -738,7 +743,7 @@ func TestStrategyUpdate(t *testing.T) {
 		},
 		"drop-fields-prioritized-list": {
-			oldObj:                obj,
+			oldObj:                objWithAdminAccessInNonAdminNamespace,
```
The only difference between `obj` and `objWithPrioritizedList` should be the addition of prioritized list fields. The namespace needs to be the same. In general, `obj` should be the "base" object for all test cases. The only objects with non-default namespace should be the objects explicitly set up to test admin access.

 I don't know when this went sideways, but let's fix it now.

sunya-ch replied:
@pohly Thank you for confirming. I fixed the namespace.

pohly commented on pkg/registry/resource/resourceclaimtemplate/strategy_test.go:
```diff
@@ -815,11 +825,15 @@ func TestStrategyUpdate(t *testing.T) {
+				assert.Len(t, errs, expectedErrLen, "exact number of errors expected")
+				for i, expectErr := range tc.expectValidationErrors {
+					assert.ErrorContains(t, errs[i], expectErr, "the error message should have contained the expected error message")
+				}
```
```suggestion
				if assert.Len(t, errs, expectedErrLen, "exact number of errors expected") {
					for i, expectErr := range tc.expectValidationErrors {
						assert.ErrorContains(t, errs[i], expectErr, "the error message should have contained the expected error message")
					}
				}
```

Don't compare the elements unless the length is the same, it would panic the test.

pohly replied:
/lgtm
/hold

Feel free to lift the hold. A panic is still going to show up as a test failures, just not as a nice one.

pohly reviewed (changes requested):
Almost there...

pohly commented:
/lgtm
/hold cancel
